### PR TITLE
[nrf fromlist] drivers/flash/flash_simulator: No real allocation...

### PR DESCRIPTION
**Pull request by de-nordic**
... just simulation

The commit adds CONFIG_FLASH_SIMULATOR_NOALLOC Kconfig option
that will cause the flash simulator driver to allocate maximum
of write_block_size, while still respecting size and offset in DTS,
and pretending to do read/write/erase operations.

Upstream PR: https://github.com/zephyrproject-rtos/zephyr/pull/38524

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

### CODE REVIEW

File: drivers/flash/flash_simulator.c [hunk 1]
```diff
@@ -51,7 +51,12 @@
 #error "Erase unit must be a multiple of program unit"
 #endif
 
+#ifndef CONFIG_FLASH_SIMULATOR_NOALLOC
 #define MOCK_FLASH(addr) (mock_flash + (addr) - FLASH_SIMULATOR_BASE_OFFSET)
+#else
+#define MOCK_FLASH(addr) (mock_flash + ((addr) % FLASH_SIMULATOR_PROG_UNIT) - \
+			  FLASH_SIMULATOR_BASE_OFFSET)
```
Comment (carlescufi) [inline drivers/flash/flash_simulator.c:58]: This might go behind the address of `mock_flash`?

File: drivers/flash/Kconfig.simulator [hunk 2]
```diff
@@ -26,6 +26,14 @@ config FLASH_SIMULATOR_DOUBLE_WRITES
 	 If selected, writing to a non-erased program unit will succeed, otherwise, it will return an error.
 	 Keep in mind that write operations can only pull bits to zero, regardless.
 
+config FLASH_SIMULATOR_NOALLOC
+	bool "No buffer allocated, reads, writes and erase are simulated"
```
Comment (MarekPieta) [inline drivers/flash/Kconfig.simulator:30]: Only buffer of write block size allocated ...

File: drivers/flash/flash_simulator.c [hunk 3]
```diff
@@ -51,7 +51,11 @@
 #error "Erase unit must be a multiple of program unit"
 #endif
 
+#ifndef CONFIG_FLASH_SIMULATOR_NOALLOC
 #define MOCK_FLASH(addr) (mock_flash + (addr) - FLASH_SIMULATOR_BASE_OFFSET)
+#else
+#define MOCK_FLASH(addr) (mock_flash + ((addr) & 0))
```
Comment (MarekPieta) [inline drivers/flash/flash_simulator.c:57]: #define MOCK_FLASH(addr) (mock_flash) ?

File: drivers/flash/Kconfig.simulator (hunk 2 above)
Comment (de-nordic) [reply drivers/flash/Kconfig.simulator:30]: OK, will fix.

File: drivers/flash/flash_simulator.c (hunk 3 above)
Comment (MarekPieta) [reply drivers/flash/flash_simulator.c:57]: Discussed offline

File: drivers/flash/flash_simulator.c (hunk 1 above)
Comment (de-nordic) [reply drivers/flash/flash_simulator.c]: Fixed.

**Review by nvlsianpu — APPROVED** [review]

File: drivers/flash/Kconfig.simulator [hunk 4]
```diff
@@ -26,6 +26,15 @@ config FLASH_SIMULATOR_DOUBLE_WRITES
 	 If selected, writing to a non-erased program unit will succeed, otherwise, it will return an error.
 	 Keep in mind that write operations can only pull bits to zero, regardless.
 
+config FLASH_SIMULATOR_NOALLOC
+	bool "Reads, writes and erase are simulated within buffer of write_block_size size"
+	depends on !NATIVE_POSIX
```
Comment (MarekPieta) [inline drivers/flash/Kconfig.simulator:31]: `!ARCH_POSIX` ?

**Review by MarekPieta — CHANGES_REQUESTED** [review]

**Review by MarekPieta — APPROVED** [review]

**Comment by de-nordic** [comment]
Implemented in other way.

**Pull request closed without merging**